wolfcrypt/src/evp.c

```diff
@@ -6803,6 +6803,833 @@ void wolfSSL_EVP_PKEY_free(WOLFSSL_EVP_PKEY* key)
         }
     }
 }
+#if defined(OPENSSL_EXTRA)
+/* Converts byte data into 2 digit hex code
+ * then put them to the specified buffer.
+ */
+static int ToHex( byte in, byte* hex )
+{
+    byte UpNibble,LwNibble;
+    byte HexTbl[16] = { '0','1','2','3','4','5','6','7',
+                        '8','9','a','b','c','d','e','f' };
+
+    if ( hex == NULL )
+        return 0;
+
+    UpNibble = (in >> 4) & 0x0f;
+    LwNibble =  in & 0x0f;
+    *hex++ = HexTbl[UpNibble];
+    *hex   = HexTbl[LwNibble];
+    return 2;
+}
+/* Converts input value "i" to upto five digit decimal
+ * and copys to the specified buffer.
+ */
+static int ToDec(word32 in, byte* hex)
+{
+    int     i = 0;
+    byte    dgt[5];
+    word32  quo;
+    int     written;
+
+    if (hex == NULL || in > 99999 )
+        return 0;
+    quo = in;
+    dgt[4] = quo % 10;
+    quo    = quo / 10;
+    dgt[3] = quo % 10;
+    quo    = quo / 10;
+    dgt[2] = quo % 10;
+    quo    = quo / 10;
+    dgt[1] = quo % 10;
+    quo    = quo / 10;
+    dgt[0] = quo % 10;
+
+    /* to remove leading zero */
+    if (dgt[0] == 0) {
+        if (dgt[1] == 0) {
+            if (dgt[2] == 0) {
+                if (dgt[3] == 0) {
+                    i = 4;
+                }else
+                    i = 3;
+            }else
+                i = 2;
+        }else
+            i = 1;
+    }else
+        i = 0;
+
+    written = 5 - i;
+
+    for (; i < 5; i++) {
+        *hex++ = dgt[i] + '0';
+    }
+    return written;
+}
+/* Indent adds white spaces of the number specified by "indents"
+ * to the buffer specified by "dst".
+ */
+static int Indent(int indents, byte* dst )
+{
+    int i;
+    if (dst == NULL)
+        return 0;
+
+    for (i = indents; i; i--) {*dst++ = ' ';}
+    return indents;
+}
+/* DumpElement dump byte-data specified by "input" to the "out".
+ * Each line has leading white spaces( "indent" gives the number ) plus
+ * four spaces, then hex coded 15 byte data with separator ":" follow.   
+ * Each line looks like: 
+ * "    00:e6:ab: --- 9f:ef:"
+ */
+static int DumpElement(WOLFSSL_BIO* out, const byte* input,
+    int inlen, int indent)
+{
+    word32 in = 0;
+    word32 i;
+    int    idx = 0;
+    int    wsz = 0;
+    byte   buff[128] = { 0 };
```

Comment:
Introduced the smallstack pattern. And reduced required buffer size to 80 bytes.

Reply:
please add a check on the return of XMALLOC to catch the case where it could be null, and I noticed that later on there is some wolfSSL_BIO_write calls in the function that should have a check on the return value

Reply:
Added return value checks.